### DISCUSSION TITLE
question/can-you-shuffle-an-array

### DISCUSSION THREAD

**Discussion by giscus** [us-fe/us-fe.github.io#20] [Announcements]
# question/can-you-shuffle-an-array

北美前端面试攻略

https://us-fe.github.io/question/can-you-shuffle-an-array.html

**Comment by xiaoyu2er** [comment]
https://javascript.info/array-methods#shuffle-an-array

经典解法

There’s a great algorithm called [Fisher-Yates shuffle](https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle). The idea is to walk the array in the reverse order and swap each element with a random one before it.

```js
function shuffle(array) {
  for (let i = array.length - 1; i > 0; i--) {
    let j = Math.floor(Math.random() * (i + 1)); // random index from 0 to i

    // swap elements array[i] and array[j]
    // we use "destructuring assignment" syntax to achieve that
    // same can be written as:
    // let t = array[i]; array[i] = array[j]; array[j] = t
    [array[i], array[j]] = [array[j], array[i]];
  }
}
```
